File: GUI/GameScene.cs
namespace prototype.GUI
{
    using Game;
    using static SDL2.SDL;

    internal class GameScene : IScene
    {
        private readonly TextRenderer textRenderer_;
        private readonly Prototype prototype_;
        private readonly Options options_;
        private static readonly SDL_Color ErrorColor = new SDL_Color { r = 0xff };
        private static readonly SDL_Color DefaultColor = new SDL_Color { r = 0xff, g = 0xff, b = 0xff };
        private SDL_Color currentColor_ = DefaultColor;

        public GameScene(Prototype prototype, TextRenderer textRenderer, Options options)
        {
            textRenderer_ = textRenderer;
            prototype_ = prototype;
            options_ = options;
        }

        public void Render()
        {
            var targetArea = CalculateRenderArea(options_.ScreenWidth, options_.ScreenHeight);
            textRenderer_.Render(prototype_.Text, targetArea, currentColor_);
        }

        public TargetSceneType HandleEvent(SDL_Event e)
        {
            if (e.type == SDL_EventType.SDL_KEYDOWN && e.key.keysym.sym == SDL_Keycode.SDLK_ESCAPE)
            {
                return TargetSceneType.MainMenu;
            }

            if (e.type == SDL_EventType.SDL_TEXTINPUT)
            {
                var character = SdlInputHelper.GetPressedCharacter(e.text);
                currentColor_ = prototype_.Type(character) ? DefaultColor : ErrorColor;
            }

            return TargetSceneType.Unchanged;
        }

        //TODO get Defaults.LetterHeight out of here.
        // perhaps give text renderer an option to render vertically centered within an area
        private static SDL_Rect CalculateRenderArea(int width, int height) =>
            new SDL_Rect
            {
                x = (int)(0.3 * width),
                y = (int)(0.5 * height - 0.5 * Defaults.LetterHeight),
                w = (int)(0.5 * width)
            };
    }
}

File: src/game_scene.py
from scene_type import SceneType
from color_scheme import ERROR_COLOR, DEFAULT_COLOR
from sdl2 import *

class GameScene(object):
    def __init__(self, text_renderer, options):
        self._text_renderer = text_renderer
        self._options = options

    def handle_event(self, event):
        if event.type == SDL_KEYDOWN and event.key.keysym.sym == SDLK_ESCAPE:
            return SceneType.MAIN_MENU
        if event.type == SDL_TEXTINPUT:
            char = event.text.text
            print(char)
            print(type(char))
            return SceneType.MAIN_MENU
        return SceneType.UNCHANGED
    
    def render(self):
        area = self._calculate_render_area()
        self._text_renderer.render("Hello guys, this is some foo bar baz", area, ERROR_COLOR)
    
    def _calculate_render_area(self):
        return SDL_Rect(
            int(0.3 * self._options.screen_width),
            int(0.5 * self._options.screen_height - 0.5 * self._options.scaled_letter_height),
            int(0.5 * self._options.screen_width),
        )